### ISSUE TITLE
Variadic function Vector creation

### ISSUE THREAD

**Issue by bmabsout** [expipiplus1/vector-sized#18]
Hi, I reinvented the wheel before I found your library. But when I wanted to create the actual vectors at compile time I was stuck with no way to make them in a general way. Until I was able to make a variadic function:
```haskell
class Make n a r | r -> a where
    make :: Sized n a -> r

instance Make n a (Sized n a) where
    make x = sreverse x
    {-# INLINE make #-}

instance (Make (n+1) a r) => Make n a (a -> r) where
    make acc a = make (a `scons` acc)
    {-# INLINE make #-}

mkN :: (Make 1 a r) => a -> r
mkN = make empty
```

sreverse and scons are just cons and reverse.

here's an example usage
```haskell
test :: Sized 10 Int
test = mkN 1 2 3 4 5 6 7 8 9 10
```
changing the size of the list will result in a compile error.
I saw example code using `V3` and `V4` which aren't generalized. I just thought maybe this can be used instead?

**Comment by expipiplus1** [comment]
This is what I'm using. This allows one to construct vectors like: 

```haskell
test :: Vector 5 Int
test = 1 :> 2 :> 3 :> 4 :> 5 :> Nil
```

```haskell
{-# LANGUAGE DataKinds       #-}
{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE TypeOperators   #-}
{-# LANGUAGE ViewPatterns    #-}

module Data.Vector.Generic.Sized.Extra
  ( module V
  , pattern Nil
  , pattern (:>)
  , uncons
  ) where

import           Control.Arrow
import qualified Data.Vector.Generic.Base  as B
import           Data.Vector.Generic.Sized as V
import           GHC.TypeLits

pattern Nil :: B.Vector v a => Vector v 0 a
pattern Nil <- _
  where Nil = V.empty

infixr 5 :>
pattern (:>) :: B.Vector v a => a -> Vector v n a -> Vector v (1 + n) a
pattern h :> t <- (uncons -> (h, t))
  where (:>) = V.cons

uncons :: B.Vector v a => Vector v (1 + n) a -> (a, Vector v n a)
uncons = V.head &&& V.tail
```

Notation borrowed from https://hackage.haskell.org/package/clash-prelude-0.11.2/docs/CLaSH-Sized-Vector.html#t:Vec

**Comment by expipiplus1** [comment]
> changing the size of the list will result in a compile error.

Does it result in a useful error :)

**Comment by bmabsout** [comment]
hmmm, haven't touched pattern synonyms but why isn't it simply `(:>) = V.cons`? Also I made the variadic function because I had vectors of length 100 and having that operator everywhere wasn't very syntactically nice, but it's fine too.

**Comment by expipiplus1** [comment]
The other part of the synonym make it possible to use the synonym to *deconstruct* vectors.

So one can make a function taking the first three elements of any vector (at least three in length) and putting them in a triple.

```haskell
take3 (a :> b :> c :> _) = (a, b,c)
```

**Comment by bmabsout** [comment]
Ah right, right. Thank you for teaching me some haskell. The error messages are not very good no, maybe it can be fixed with haskell's new custom error support? Again I am new to this so I haven't tried it yet. Maybe we should close this issue since the improvement is silly.

**Comment by expipiplus1** [comment]
Ok, thanks very much for the idea. If you get a version working at some point which gives really good error messages we should revisit this!

**Comment by bmabsout** [comment]
Alright I added a good error message, turns out it's actually pretty simple.
```Haskell
class Make n a r | r -> a where
    make :: Sized n a -> r

instance (ErrorIfUnequal n x, n~x) => Make n a (Sized x a) where
    make x = sreverse x
    {-# INLINE make #-}

instance (Make (n+1) a r) => Make n a (a -> r) where
    make acc a = make (a `scons` acc)
    {-# INLINE make #-}

type family ErrorIfUnequal (a :: Nat) (b :: Nat) where
  ErrorIfUnequal a a = (() :: K.Constraint)
  ErrorIfUnequal a b = TypeError (Text "The created vector has size: " :<>: ShowType a :$$: Text " While the expected size is: " :<>: ShowType b)

mkN :: (Make 1 a r) => a -> r
mkN = make empty
```
trying this: 
```Haskell
test :: Sized 3 Int
test = mkN 1 2
```
gives the error :
```
    • The created vector has size: 2
       While the expected size is: 3
    • In the expression: mkN 1 2
      In an equation for ‘test’: test = mkN 1 2
```

**Comment by bmabsout** [comment]
Actually there is one more benefit here although I doubt anyone will have this problem. Since creating a vector the way you showed me conses every time, the operation will be O(n^2). You can use this approach to cons a list and then fromList to create the vector so it is O(n).

**Comment by expipiplus1** [comment]
Cool, nice error.

What kind of inference properties does this have.

What do these output?

```haskell
test0 :: _
test0 = mkN

test3 :: _
test3 = mkN 1 2 3

testIncorrect :: _
testIncorrect = mkN True () 0
```

**Comment by bmabsout** [comment]
With `PartialSignatures` off, all 3 produce an ambiguous type variable exception and with not very useful errors.
With it on you get these:
```haskell
test0 :: _
test0 = mkN
```
```
/home/bmabsout/Documents/neural-swarm/src/SizedL.hs:153:9: error:
    • Ambiguous type variables ‘a0’, ‘r0’ arising from a use of ‘mkN’
      prevents the constraint ‘(Make 1 a0 r0)’ from being solved.
      Relevant bindings include
        test0 :: a0 -> r0
          (bound at /home/bmabsout/Documents/neural-swarm/src/SizedL.hs:153:1)
      Probable fix: use a type annotation to specify what ‘a0’, ‘r0’ should be.
      These potential instances exist:
        instance (ErrorIfUnequal n x, n ~ x) => Make n a (Sized x a)
          -- Defined at /home/bmabsout/Documents/neural-swarm/src/SizedL.hs:136:10
        instance Make (n + 1) a r => Make n a (a -> r)
          -- Defined at /home/bmabsout/Documents/neural-swarm/src/SizedL.hs:140:10
    • In the expression: mkN
      In an equation for ‘test0’: test0 = mkN
```
```haskell
test3 :: _
test3 = mkN 1 2 3
```
```
/home/bmabsout/Documents/neural-swarm/src/SizedL.hs:156:9: error:
    • Ambiguous type variables ‘a1’, ‘t0’ arising from a use of ‘mkN’
      prevents the constraint ‘(Make 3 a1 t0)’ from being solved.
      Relevant bindings include
        test3 :: t0
          (bound at /home/bmabsout/Documents/neural-swarm/src/SizedL.hs:156:1)
      Probable fix: use a type annotation to specify what ‘a1’, ‘t0’ should be.
      These potential instances exist:
        instance (ErrorIfUnequal n x, n ~ x) => Make n a (Sized x a)
          -- Defined at /home/bmabsout/Documents/neural-swarm/src/SizedL.hs:136:10
        instance Make (n + 1) a r => Make n a (a -> r)
          -- Defined at /home/bmabsout/Documents/neural-swarm/src/SizedL.hs:140:10
    • In the expression: mkN 1 2 3
      In an equation for ‘test3’: test3 = mkN 1 2 3
```
```
/home/bmabsout/Documents/neural-swarm/src/SizedL.hs:156:13: error:
    • Ambiguous type variable ‘a1’ arising from the literal ‘1’
      prevents the constraint ‘(Num a1)’ from being solved.
      Probable fix: use a type annotation to specify what ‘a1’ should be.
      These potential instances exist:
        instance Num Integer -- Defined in ‘GHC.Num’
        instance Num a => Num (Vec a)
          -- Defined at /home/bmabsout/Documents/neural-swarm/src/Convenience.hs:134:10
        instance Num Double -- Defined in ‘GHC.Float’
        ...plus three others
        ...plus 53 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    • In the first argument of ‘mkN’, namely ‘1’
      In the expression: mkN 1 2 3
      In an equation for ‘test3’: test3 = mkN 1 2 3
```
```haskell
testIncorrect :: _
testIncorrect = mkN True () 0
```
```
/home/bmabsout/Documents/neural-swarm/src/SizedL.hs:159:17: error:
    • Couldn't match type ‘()’ with ‘Bool’
        arising from a functional dependency between:
          constraint ‘Make 1 Bool (() -> t1 -> t)’
            arising from a use of ‘mkN’
          instance ‘Make n a (a -> r)’
            at /home/bmabsout/Documents/neural-swarm/src/SizedL.hs:140:10-46
    • In the expression: mkN True () 0
      In an equation for ‘testIncorrect’: testIncorrect = mkN True () 0
```

**Comment by mstksg** [comment]
Any reason why this wasn't pursued further? it could be kind of useful :)

**Comment by expipiplus1** [comment]
Sorry for the delayed response.

@bmabsout I think this could be a nice addition to this package, provided that there's a nice bit of documentation making it clear when the type inference isn't the best. Please open a PR! :)

**Comment by bmabsout** [comment]
woops, I wasn't receiving emails associated to this github account. I didn't end up improving it (I tried a bit) though since I use `PartialTypeSignatures` and adding `:: Sized _ _` is not that costly. Maybe someone who understands these better than me can fix it? but I did do something else involving type lists:
```haskell
{-# LANGUAGE PolyKinds #-}
type family Length (a :: [k]) = (n::Nat) where
  Length '[] = 0
  Length (x ': xs) = 1 + Length xs

class  MakeVector list where
  create :: Num a => Sized (Length list) a
instance MakeVector '[] where
  create = empty

instance forall x xs. (KnownNat x, MakeVector xs) => MakeVector ((x::Nat) ': xs) where
  create = fromIntegral (natVal (Proxy :: Proxy x)) `scons` create @xs

test = create @[1,23,2,4]
```
`test :: Num a => Sized 4 a`

`Sized n a` is the list with type level size
`empty` creates a `Sized 0 a`
`scons` is just cons so of type (a -> Sized n a -> Sized (n+1) a)
this requires allowing ambiguous types because Length isn't injective, otherwise a Proxy could be added and it would be used like this `create (Proxy @(Proxy [1,2,3,4]))`
problem with it is that it only supports Nats which is horrible. Maybe someone who knows more can also make this more general? not sure how to promote any literal and then demote it. I like this more because it is less hacky than variadic functions
